### PR TITLE
Work/self aware tasks

### PR DESCRIPTION
**Pull request by deathowl**
This is a patch for issue https://github.com/coleifer/huey/issues/93.
Works the following way:
if the function has a self parameter, the task instance is passed to it, so we can do pretty fancy stuff inside the executed task. For example report progess to the pubsub 
```
@long_running_huey.task()
def run_long(a,b,c,self):
    for i in range(1, 100):
        progress  = {"progress": i}
        long_running_huey.emit_task(progress, self)
        sleep(1)
```

### CODE REVIEW

**Comment by coleifer** [comment]
Interesting approach. I would prefer, though, to follow the "retries_as_parameter" approach and add an additional param to the `@task` decorator itself for including the instance, e.g.

```python
@huey.task(include_task=True)
def my_task(foo, bar, task=None):
    do_something_with_task_obj()
```

**Comment by deathowl** [comment]
So i'll do as you prefer, expect changed patch soon.

File: huey/api.py [hunk 1]
```diff
@@ -6,7 +6,6 @@
 import traceback
 import uuid
 from functools import wraps
-
```
Comment (coleifer) [inline huey/api.py]: Why is this removed?

File: huey/api.py [hunk 2]
```diff
@@ -45,7 +44,6 @@ class Huey(object):
 
         from huey.api import Huey, crontab
         from huey.backends.redis_backend import RedisQueue, RedisDataStore, RedisSchedule
-
```
Comment (coleifer) [inline huey/api.py]: why is this removed?

File: huey/tests/queue.py [hunk 3]
```diff
@@ -228,6 +235,21 @@ def test_execution(self):
...
 
+    def test_self_awareness(self):
+        self_aware('k', 'v')
+        task = huey.dequeue()
+        huey.execute(task)
+        self.assertEqual(last_executed_task_class.pop(), "queuecmd_self_aware")
+
+        self_aware('k', 'v')
+        huey.execute(huey.dequeue())
+        self.assertEqual(last_executed_task_class.pop(), "queuecmd_self_aware")
```
Comment (coleifer) [inline huey/tests/queue.py]: Any reason this is called twice? Seems to be doing the same thing as the previous block.

**Comment by coleifer** [comment]
Looks great, just a couple minor things and I'll merge it up. Thanks for this.

File: huey/tests/queue.py (hunk 3 above)
Comment (deathowl) [inline huey/tests/queue.py]: i just followd the structure of your tests.

File: huey/api.py [hunk 4]
```diff
@@ -74,13 +72,13 @@ def __init__(self, queue, result_store=None, schedule=None, events=None,
         self.store_none = store_none
         self.always_eager = always_eager
 
-    def task(self, retries=0, retry_delay=0, retries_as_argument=False,
+    def task(self, retries=0, retry_delay=0, retries_as_argument=False, include_task=False,
```
Comment (coleifer) [inline huey/api.py]: Just some more notes... pep8 should be 79 chars long.

File: huey/api.py [hunk 5]
```diff
@@ -74,13 +72,13 @@ def __init__(self, queue, result_store=None, schedule=None, events=None,
...
              name=None):
         def decorator(func):
             """
             Decorator to execute a function out-of-band via the consumer.
             """
-            klass = create_task(QueueTask, func, retries_as_argument, name)
+            klass = create_task(QueueTask, func, retries_as_argument, name, include_task)
```
Comment (coleifer) [inline huey/api.py]: ditto here line length.

File: huey/api.py [hunk 6]
```diff
@@ -409,12 +407,14 @@ def validate_datetime(self, dt):
         return False
 
 
-def create_task(task_class, func, retries_as_argument=False, task_name=None,
+def create_task(task_class, func, retries_as_argument=False, task_name=None, include_task=False,
```
Comment (coleifer) [inline huey/api.py]: ditto here line length.

**Comment by laike9m** [comment]
Could you provide an example on how to use self aware task? I'm not sure what properties/information task object can offer. I would like to know how many times this task has been executed(retried), is this possible?

**Comment by deathowl** [comment]
An example:
`__author__ = 'deathowl'

from time import sleep

from jobqueue.config import simple_huey


@simple_huey.task(include_task=True)
def simple_task(task=None, time=100000):
    """
    :param msg: Message
    """
    from utils.globals import app
    with app.app_context():
        for i in range(1, time):
            simple_huey.emit_task(i, task)
            sleep(1)
    return "done"
`

**Comment by laike9m** [comment]
@deathowl Thank yuo. Can I get the information of how many times this task has been executed(retried) from task object?
